source/slang/slang-emit-spirv.cpp

```diff
@@ -0,0 +1,1141 @@
+        // > all contents past the end of the string in the final word are padded with 0.
+        //
+        // For the last word, the low-order bytes will
+        // come from the remainder of the string (if
+        // there is anything left), and the rest will
+        // be left as zeros.
+        //
+        // TODO: This code should probably assert that `text`
+        // doesn't contain any embedded nul bytes, since they
+        // could lead to invalid encoded results.
+        //
+        SLANG_ASSERT((end - cursor) <= 3);
+        SpvWord lastWord = 0;
+        memcpy(&lastWord, cursor, (end - cursor));
+        emitOperand(dst, lastWord);
+    }
+
+    // Sometimes we will want to pass down an argument that
+    // represents a result <id> operand, but we won't yet
+    // have access to the `SpvInst` that will get the <id>.
+    // We will use a dummy `enum` type to support this case.
+
+    enum ResultIDToken { kResultID };
+
+    void emitOperand(SpvInst* dst, ResultIDToken)
+    {
+        // A result <id> operand uses the <id> of the instruction itself.
+        emitOperand(dst, getID(dst));
+    }
+
+    // As another convenience, there are often cases where
+    // we will want to emit all of the operands of some
+    // IR instruction as <id> operands of a SPIR-V
+    // instruction. This is handy in cases where the
+    // Slang IR and SPIR-V instructions agree on the
+    // number, order, and meaning of their operands.
+
+        /// Helper type for emitting all the operands of some IR instruction
+    struct OperandsOf
+    {
+        OperandsOf(IRInst* irInst)
+            : irInst(irInst)
+        {}
+
+        IRInst* irInst = nullptr;
+    };
+
+        /// Emit operand words for all the operands of a given IR instruction
+    void emitOperand(SpvInst* dst, OperandsOf const& other)
+    {
+        auto irInst = other.irInst;
+        auto operandCount = irInst->getOperandCount();
+        for( UInt ii = 0; ii < operandCount; ++ii )
+        {
+            emitOperand(dst, irInst->getOperand(ii));
+        }
+    }
+
+    // With the above routines, code can easily construct a SPIR-V
+    // instruction with arbitrary operands over multiple lines of code.
+    //
+    // In many cases, however, it is desirable to be able to emit
+    // an instruction more compactly, and for that we will introduce
+    // a number of `emitInst()` helpers that handle creating an
+    // instruction, filling in its operands, and adding it to a parent.
+    //
+    // These routines are overloaded on the number of operands, and
+    // also templates to work with any of the types for which
+    // `emitOperand()` works.
+    //
+    // In all of these cases, the caller takes responsibility for
+    // correctly matching the SPIR-V encoding rules for the chosen
+    // opcode, including whether a type <id> or result <id> is
+    // required.
+
+    SpvInst* emitInst(SpvInstParent* parent, IRInst* irInst, SpvOp opcode)
```

Comment:
I can see the point, but the idea here is that the function is `emitInst` and `irInst` *is* the instruction being emitted, so it seems to be a good one to have first, while the `opcode` is really just a detail of `irInst`.

(The fact that `parent` comes first is just because in an ideal world this would be `parent->emitInst(...))

Another, subtle, reason for putting `irInst` before the opcode is that it makes a clear separation that everything after the SPIR-V opcode represents a SPIR-V operand. If we had the `SpvOp` come earlier then in a call like `emitInst(parent, spvOp, a, b, c, d)` you might have all of `a, b, c, d, ...` be `IRInst`s, and it could be tricky to realize at the call site that the meaning of `a` there is completely different than `b, c, d`.

(It also means we could in theory make the `irInst` operand optional in the future, since it is allowed to be `nullptr` right now).